[Inductor] GEMM Shape Padding Optimization

jiawenliu64 opened the pull request:
Summary:
Optimize the shape padding in the following perspectives:
- Add BFloat16 support for AMP training and Float16 support for inference
- Optimize microbenchmark to avoid peak memory issue, and include profiling memory ops to make more accurate decision
- Add a flag to turn off/on padding dims N and M in `torch.bmm` due to expensive memory copy of `.contiguous` to avoid peak memory issues in internal models

Test Plan: CI

Differential Revision: D41724868



cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

jianyuh commented on torch/_inductor/decomposition.py:
```diff
@@ -243,85 +251,103 @@ def should_pad_bench(mat1, mat2, op, input=None):
-            return ori_time > pad_time * 1.3
-        else:
-            return ori_time > pad_time * 2
+        return ori_time > pad_time * 1.1
```
Keep the comment on how 1.1 is chosen?

ngimel commented on torch/_inductor/decomposition.py:
```diff
@@ -243,85 +251,105 @@ def should_pad_bench(mat1, mat2, op, input=None):
-        k_padded_length = get_padded_length(mat1.shape[2])
-        n_padded_length = get_padded_length(mat2.shape[2])
-
+        k_padded_length = get_padded_length(mat1.shape[2], get_alignment_size(mat1))
```
In multihead attention with odd sequence length, seq len dimension is m and n, and padding actually helps a lot, can we leave m and n padding? Where are you seeing `.contiguous` call with padded m and n?

jiawenliu64 replied:
Great to see shape padding helps multihead attention a lot! We observed it in two internal production models. 

Probably we could add one flag to turn off/on padding m and n dims of bmm for internal models, and turn on by default for multihead attention when enabling shape padding?

jiawenliu64 replied:
When we tested it in 128 GPUs with delay-allreduce, there was an OOM issue due to padding m and n of bmm for two internal models, though the OOM issue does not happen in single GPU

bertmaher replied:
I kind of persist in being skeptical that the IMA we saw in the 128-gpu case was really caused by OOMs here.  Even if it was, we should leave some kind of a TODO or whatever to make it so we can turn this padding on (maybe with more memory analysis?) since we're kind of leaving perf on the table.

jiawenliu64 replied:
Thanks @bertmaher! I will follow up on this with more analysis, and will further remove the flag once the OOM issue is addressed. Will keep you updated

Commit:
[Inductor] GEMM Shape Padding Optimization (#90425)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/90425

Optimize the shape padding in the following perspectives:
- Add BFloat16 support for AMP training and Float16 support for inference
- Optimize microbenchmark to avoid peak memory issue, and include profiling memory ops to make more accurate decision
- Turn off padding dims N and M in `torch.bmm` due to expensive memory copy of `.contiguous` to avoid peak memory issue

Test Plan: CI

Reviewed By: jianyuh

Differential Revision: D41724868

fbshipit-source-id: 858f6ef8845a61e3aa29504b067eab676fce46f9

jiawenliu64 commented:
@pytorchbot merge
